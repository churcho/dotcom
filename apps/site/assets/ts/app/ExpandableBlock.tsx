import React, { ReactElement, useState } from "react";
import { handleReactEnterKeyPress } from "../helpers/keyboard-events";
import renderSvg from "../helpers/render-svg";

export interface ExpandableBlockHeader {
  text: string;
  iconSvgText: string | null;
}

interface Props {
  initiallyExpanded: boolean;
  header: ExpandableBlockHeader;
  children: ReactElement<HTMLElement>;
  id: string;
}

const caret = (expanded: boolean): ReactElement<HTMLElement> => {
  const unicodeCharacter = expanded ? "&#xF107;" : "&#xF106;";
  return (
    <span
      className="c-expandable-block__header-caret"
      // eslint-disable-next-line react/no-danger
      dangerouslySetInnerHTML={{ __html: unicodeCharacter }}
    />
  );
};

export default ({
  initiallyExpanded,
  header: { text, iconSvgText },
  children,
  id
}: Props): ReactElement<HTMLElement> => {
  const [expanded, toggleExpanded] = useState(initiallyExpanded);
  const headerId = `header-${id}`;
  const panelId = `panel-${id}`;

  const onClick = (): void => toggleExpanded(!expanded);

  return (
    <>
      <div
        className="c-expandable-block__header"
        tabIndex={0}
        id={headerId}
        aria-expanded={expanded}
        aria-controls={panelId}
        role="button"
        onClick={onClick}
        onKeyPress={e => handleReactEnterKeyPress(e, onClick)}
      >
        {iconSvgText
          ? renderSvg("c-expandable-block__header-icon", iconSvgText)
          : null}
        {text}
        {caret(expanded)}
      </div>
      {expanded ? (
        <div
          className="c-expandable-block__panel"
          // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
          tabIndex={0}
          role="region"
          id={panelId}
          aria-labelledby={headerId}
          ref={panel => panel && panel.focus()}
        >
          {children}
        </div>
      ) : null}
    </>
  );
};
