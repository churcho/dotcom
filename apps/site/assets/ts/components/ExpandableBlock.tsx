import React, { ReactElement, useState } from "react";
import { handleReactEnterKeyPress } from "../helpers/keyboard-events";
import renderSvg from "../helpers/render-svg";
import { caret } from "../helpers/icon";

export interface ExpandableBlockHeader {
  text: string;
  iconSvgText: string | null;
}

interface Props {
  initiallyExpanded: boolean;
  header: ExpandableBlockHeader;
  children: ReactElement<HTMLElement>;
  id: string;
}

interface State {
  expanded: boolean;
  focused: boolean;
}

export default ({
  initiallyExpanded,
  header: { text, iconSvgText },
  children,
  id
}: Props): ReactElement<HTMLElement> => {
  const [state, toggleExpanded] = useState({
    expanded: initiallyExpanded,
    focused: false
  });
  const { expanded, focused }: State = state;
  const headerId = `header-${id}`;
  const panelId = `panel-${id}`;

  const onClick = (): void =>
    toggleExpanded({ expanded: !expanded, focused: true });

  return (
    <>
      <h3
        className="c-expandable-block__header"
        tabIndex={0}
        id={headerId}
        aria-expanded={expanded}
        aria-controls={panelId}
        // eslint-disable-next-line jsx-a11y/no-noninteractive-element-to-interactive-role
        role="button"
        onClick={onClick}
        onKeyPress={e => handleReactEnterKeyPress(e, onClick)}
      >
        {iconSvgText
          ? renderSvg("c-expandable-block__header-icon", iconSvgText)
          : null}
        {text}
        {caret("c-expandable-block__header-caret", expanded)}
      </h3>
      {expanded ? (
        <div
          className="c-expandable-block__panel"
          // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
          tabIndex={0}
          role="region"
          id={panelId}
          aria-labelledby={headerId}
          ref={panel => panel && focused && panel.focus()}
        >
          {children}
        </div>
      ) : null}
      {/* No javascript support */}
      <noscript>
        <style>{`#${headerId} { display: none; }`}</style>
        <h3 className="c-expandable-block__header">
          {iconSvgText
            ? renderSvg("c-expandable-block__header-icon", iconSvgText)
            : null}
          {text}
          {caret("c-expandable-block__header-caret", true)}
        </h3>
        <div className="c-expandable-block__panel" role="region">
          {children}
        </div>
      </noscript>
    </>
  );
};
