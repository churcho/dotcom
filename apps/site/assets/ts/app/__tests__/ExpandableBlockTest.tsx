import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import ExpandableBlock from "../ExpandableBlock";
// @ts-ignore
import accessibleIcon from "../../../static/images/icon-accessible-default.svg";

const body = '<div id="react-root"></div>';

test("render expandable block expanded by default", () => {
  document.body.innerHTML = body;

  const expandableComponent = (
    <ExpandableBlock
      initiallyExpanded
      id="accessibility"
      header={{
        text: "Accessibility",
        iconSvgText: accessibleIcon
      }}
    >
      <div>
        <p>South Station is accessible. It has the following features:</p>
        <p>This is a test</p>
      </div>
    </ExpandableBlock>
  );

  const tree = renderer.create(expandableComponent).toJSON();
  expect(tree).toMatchSnapshot();
});

test("handle click to expand and enter to collapse", () => {
  document.body.innerHTML = body;

  const wrapper = mount(
    <ExpandableBlock
      initiallyExpanded={false}
      id="accessibility"
      header={{
        text: "Accessibility",
        iconSvgText: accessibleIcon
      }}
    >
      <div>
        <p>South Station is accessible. It has the following features:</p>
        <p>This is a test</p>
      </div>
    </ExpandableBlock>
  );

  expect(
    wrapper.find(".c-expandable-block__header").prop("aria-expanded")
  ).toEqual(false);

  wrapper.find(".c-expandable-block__header").simulate("click");

  expect(
    wrapper.find(".c-expandable-block__header").prop("aria-expanded")
  ).toEqual(true);

  wrapper
    .find(".c-expandable-block__header")
    .simulate("keypress", { key: "Enter" });

  expect(
    wrapper.find(".c-expandable-block__header").prop("aria-expanded")
  ).toEqual(false);
});
