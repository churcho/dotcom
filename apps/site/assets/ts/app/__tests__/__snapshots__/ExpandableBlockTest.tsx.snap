// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render expandable block expanded by default 1`] = `
Array [
  <div
    aria-controls="panel-accessibility"
    aria-expanded={true}
    className="c-expandable-block__header"
    id="header-accessibility"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span
      className="c-expandable-block__header-icon"
      dangerouslySetInnerHTML={
        Object {
          "__html": "SVG",
        }
      }
    />
    Accessibility
    <span
      className="c-expandable-block__header-caret"
      dangerouslySetInnerHTML={
        Object {
          "__html": "&#xF107;",
        }
      }
    />
  </div>,
  <div
    aria-labelledby="header-accessibility"
    className="c-expandable-block__panel"
    id="panel-accessibility"
    role="region"
    tabIndex={0}
  >
    <div>
      <p>
        South Station is accessible. It has the following features:
      </p>
      <p>
        This is a test
      </p>
    </div>
  </div>,
]
`;
